common/src/chain/transaction/signature.rs

```diff
@@ -0,0 +1,254 @@
+    hash_encoded_to(&u32::MAX, &mut stream);
+
+    let result = stream.finalize().into();
+    Ok(result)
+}
+
+fn verify_standard_input_signature(
+    outpoint_destination: &Destination,
+    witness: &StandardInputSignature,
+    tx: &Transaction,
+    input_num: usize,
+) -> Result<(), TransactionSigError> {
+    let sighash = signature_hash(witness.sighash_type(), tx, input_num)?;
+    witness.verify_signature(outpoint_destination, &sighash)?;
+    Ok(())
+}
+
+pub fn verify_signature(
+    outpoint_destination: &Destination,
+    tx: &Transaction,
+    input_num: usize,
+) -> Result<(), TransactionSigError> {
+    let target_input = tx
+        .get_inputs()
+        .get(input_num)
+        .ok_or_else(|| TransactionSigError::InvalidInputIndex(input_num, tx.get_inputs().len()))?;
+    let input_witness = target_input.get_witness();
+    match input_witness {
+        inputsig::InputWitness::NoSignature(_) => {
+            return Err(TransactionSigError::SignatureNotFound)
+        }
+        inputsig::InputWitness::Standard(witness) => {
+            verify_standard_input_signature(outpoint_destination, witness, tx, input_num)?
+        }
+    }
+    Ok(())
+}
+// TODO: write tests
+
+#[cfg(test)]
+mod test {
+    use std::vec;
+
+    use crypto::key::{KeyKind, PrivateKey};
+
+    use super::{
+        inputsig::{InputWitness, StandardInputSignature},
+        sighashtype::SigHashType,
+    };
+    use crate::{
+        chain::{
+            signature::verify_signature, Destination, Transaction, TransactionCreationError,
+            TxInput, TxOutput,
+        },
+        primitives::{Amount, Id, H256},
+    };
+
+    fn generate_tx(
+        private_key: PrivateKey,
+        sighash_type: SigHashType,
+        dest: Destination,
+    ) -> Result<Transaction, TransactionCreationError> {
+        let mut tx = Transaction::new(
+            0,
+            vec![
+                TxInput::new(
+                    Id::<Transaction>::new(&H256::zero()).into(),
+                    0,
+                    InputWitness::NoSignature(None),
+                ),
+                TxInput::new(
+                    Id::<Transaction>::new(&H256::random()).into(),
+                    1,
+                    InputWitness::NoSignature(None),
+                ),
+            ],
+            vec![TxOutput::new(Amount::from_atoms(100), dest.clone())],
+            0,
+        )?;
+
+        let signature = StandardInputSignature::produce_signature_for_input(
+            &private_key,
+            sighash_type,
+            dest.clone(),
+            &tx,
+            0,
+        )
+        .unwrap();
+
+        tx.update_witness(0, InputWitness::Standard(signature.clone())).unwrap();
+        Ok(tx)
+    }
+
+    fn sign_tx(
+        tx: &mut Transaction,
+        private_key: &PrivateKey,
+        sighash_type: SigHashType,
+        dest: Destination,
+    ) {
+        for i in 0..tx.get_inputs().len() {
```

Comment:
Let's stop using `len()` for loops to avoid off-by-one errors, and start always using `.iter().enumerate()`

Reply:
If we use `.iter().enumerate()` here, then hidden immutable borrow occurs here for `tx`. Meanwhile, in cycle, we update `tx` that occurs mutable borrow. So, I'm not entirely sure how here possible to use enumerate.

Reply:
I see. That won't work. Ignore this for now.